include/mumble/Base64.hpp

```diff
@@ -3,30 +3,147 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
+// Inspired by https://renenyffenegger.ch/notes/development/Base64/Encoding-and-decoding-base-64-with-cpp.
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		if (in.empty()) {
+			return 0;
+		}
+
+		if (in.size() % 4) {
+			throw std::runtime_error("Input is not a multiple of 4!");
+		}
+
+		const auto size = decodedSize(in);
+		if (out.size() < size) {
+			throw std::runtime_error("Insufficient output buffer size!");
+		}
+
+		for (auto data = reinterpret_cast< uint8_t * >(out.data()); !in.empty(); data += 3, in = in.subspan(4)) {
```

Comment:
Why not simply use a regular for-loop that does
```cpp
for (std::size_t i = 0; i < in.size(); ++i)
```
To me this would still seem like the most readable solution :thinking: 
All offsets/sizes can then be calculated based on the value of `i`

Reply:
It's prone to errors, also consider that we are advancing two separate iterators by a specific amount of bytes and then checking the remaining size in the code.